src/centralserver/admin-service/src/main/java/org/niis/xroad/centralserver/restapi/openapi/ClientsApiController.java

```diff
@@ -0,0 +1,136 @@
+import org.springframework.data.domain.PageRequest;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Controller
+@RequestMapping(ControllerUtil.API_V1_PREFIX)
+@PreAuthorize("denyAll")
+@RequiredArgsConstructor
+public class ClientsApiController implements ClientsApi {
+
+    private final ClientSearchService clientSearchService;
+    private final PagedClientsConverter pagedClientsConverter;
+    private final PageRequestConverter pageRequestConverter;
+    private final SecurityServerIdConverter securityServerIdConverter;
+    private final SecurityServerService securityServerService;
+
+    @Override
+    public ResponseEntity<Client> addClient(Client client) {
+        return null;
+    }
+
+    @Override
+    public ResponseEntity<Void> deleteClient(String id) {
+        return null;
+    }
+
+    @Override
+    @PreAuthorize("hasAuthority('SEARCH_MEMBERS') or hasAuthority('VIEW_MEMBERS')")
+    public ResponseEntity<PagedClients> findClients(String q,
+            PagingSortingParameters pagingSorting, String name,
+            String instance, String memberClass,
+            String memberCode, String subsystemCode,
+            ClientType clientType, String encodedSecurityServerId) {
+        PageRequest pageRequest = pageRequestConverter.convert(pagingSorting, new MemberSortParameterConverter());
+        var params = new FlattenedSecurityServerClientRepository.SearchParameters()
+                .setMultifieldSearch(q)
+                .setInstanceSearch(instance)
+                .setMemberNameSearch(name)
+                .setMemberClassSearch(memberClass)
+                .setMemberCodeSearch(memberCode)
+                .setSubsystemCodeSearch(subsystemCode)
+                .setClientType(ClientTypeMapping.map(clientType).orElse(null));
+        if (!StringUtils.isEmpty(encodedSecurityServerId)) {
+            SecurityServerId id = securityServerIdConverter.convertId(encodedSecurityServerId);
+            var securityServer = securityServerService.find(id);
+            if (securityServer.isEmpty()) {
+                throw new BadRequestException("Security server does not exist");
+            }
+            params.setSecurityServerId(securityServer.get().getId());
+        }
+        Page<FlattenedSecurityServerClientDto> page = clientSearchService.find(params, pageRequest);
+        PagedClients pagedResults = pagedClientsConverter.convert(page, pagingSorting);
+        return ResponseEntity.ok(pagedResults);
+    }
+
+    private class MemberSortParameterConverter implements PageRequestConverter.SortParameterConverter {
+        Map<String, String> conversions = new HashMap<>();
+        {
+            conversions.put("id", "id");
+            conversions.put("member_name", "memberName");
```

Comment:
It seems key and value are exchanged here. Also memberName appears twice, probably a bug?

Reply:
`            conversions.put("xroad_id.instance_id", "memberName");`

is indeed a bug, good catch!

I think key and value are in correct order though? In case of e.g. sort by member name, API call should have `sort=member_name`. Converter translates from openapi property names to entity propery names (`memberName`) since that's what the service layer uses.

Reply:
Bug fixed in https://github.com/nordic-institute/X-Road/commit/07f5424228dc92fcf9f041c5294d54ee3f5aa088